build: macOS arm and intel cross compilation.

galad87 opened the pull request:
cross compiling mac intel to mac arm and mac arm to mac intel.

**Test on:**

- [ ] Windows 10+  (via MinGW)
- [x] macOS 10.13+
- [ ] Ubuntu Linux

galad87 commented:
Here's how to create an universal build:

mkdir build
cd build
../configure
make ub

ub will resign both HandBrake.app and HandBrakeCLI, macOS on Apple Silicon needs a valid signature even if it's self-signed.

Nomis101 commented:
This does not seem to work for me. I'm getting `FAILURE (code 2)`.

[output.txt](https://github.com/HandBrake/HandBrake/files/5358550/output.txt)

galad87 commented:
Xcode 12.2 is required.
Make sure you are in the "build" folder too, the ub target seems hardcoded to use it.

If you want to build only the arm version, you can use the --xcode-config osx1011.arm64 option or  --arch arm64 --disable-xcode to compile only the cli.

Nomis101 commented:
OK, I've updated from Xcode 12.0 to 12.2 beta 2 (12B5025f). But I still get the same error. I'm on Big Sur 11.0 Beta (20A5384c), which seems the most recent beta. I am also in the "build" folder. Are there any other requirements?

[output2.txt](https://github.com/HandBrake/HandBrake/files/5358694/output2.txt)


Building for arm64 only failed with:

> checking whether make sets $(MAKE)... yes
checking whether to enable maintainer-specific portions of Makefiles... yes
checking build system type... x86_64-apple-darwin20.1.0
checking host system type... Invalid configuration `arm64-apple-darwin20.1.0': machine `arm64-apple' not recognized
configure: error: /bin/sh ./config.sub arm64-apple-darwin20.1.0 failed
make: *** [contrib/libopus/.stamp.libopus.configure] Error 1

galad87 commented:
Hopefully fixed with the latest commit.

Nomis101 commented:
Sorry, still all the same. I found out the following:

1. It seems your libopus patch does not stick. If I inspect `/build/contrib/libopus/opus-1.3.1/config.sub` the changes from the patch are not in the file. But even if I modify /download/opus-1.3.1.tar.gz with a config.sub that includes your changes, this changes are not in `/build/contrib/libopus/opus-1.3.1/config.sub`

2. It only builds in arm64 if I use `--arch=arm64 `. If I use `--xcode-config=osx1011.arm64`, it builds in x86_64.

galad87 commented:
What's the output of xcode-select -p ? Do you have only 12.2 installed of other versions too?

Nomis101 commented:
> What's the output of xcode-select -p ? Do you have only 12.2 installed of other versions too?

```
% xcode-select -p
/Applications/Xcode-beta.app/Contents/Developer
```

Yes, I have 12.2 installed only.

galad87 commented:
Do you have the command line tools in /Library/Developer too? Try to delete them.

Nomis101 commented:
> Do you have the command line tools in /Library/Developer too? Try to delete them.

No, I don't have them. This had some bad side effects, so I've deleted them long time ago.

Nomis101 commented:
I found out, why the changes for opus doesn't stick in `build/contrib/libopus/opus-1.3.1/config.sub`. This is because `autoreconf -fiv` from here https://github.com/HandBrake/HandBrake/blob/d5a58c985dcb0a9e3326a724b567c8fc1dcb6cb4/contrib/libopus/module.defs#L15 does remove the changes, after the file was patched.

By the why, do we need this changes?
https://github.com/xiph/opus/pull/187

Nomis101 commented:
I was able to successfully cross compile HandBrakeCLI for arm64.

I don't know if this is the way to go, but to do so, I first removed the force argument from autoreconf:
https://github.com/HandBrake/HandBrake/blob/d5a58c985dcb0a9e3326a724b567c8fc1dcb6cb4/contrib/libopus/module.defs#L15
`autoreconf -iv;` then the darwin-applesi-cross.patch applied nicely. 
It then stoped at libogg. So I copied the darwin-applesi-cross.patch into liboggs contrib folder and also removed the -f argument from liboggs autoreconf. The same thing I also had to do for fribidi, libvorbis, libtheora and jansson. After that the build succeeded.

```
% lipo -archs /Volumes/Developer/UB/HandBrake/build/HandBrakeCLI
arm64
```

But, `make ub.build.serial` still doesn't work.

galad87 commented:
Yes the libopus patch is useless. Still, it's quite weird because I can compile the UB on my Intel Mac… can you post the whole arm64 xcodebuild log? Which autoconf version are you using?

sr55 commented:
[build.txt](https://github.com/HandBrake/HandBrake/files/5362032/build.txt)
[config.info.txt](https://github.com/HandBrake/HandBrake/files/5362033/config.info.txt)

[config.verbose.txt](https://github.com/HandBrake/HandBrake/files/5362034/config.verbose.txt)

Building x86_64 was fine.  arm64 failed.

EDIT: Might be missing command line tools 12.2

sr55 commented:
I have a feeling it's defaulting to xcode 12.0 not 12.2

Edit: Quick setting change and rebuilding with 12.2 tools

Edit2: Failed again.
[xcodemake.env.txt](https://github.com/HandBrake/HandBrake/files/5362113/xcodemake.env.txt)
[config.verbose.txt](https://github.com/HandBrake/HandBrake/files/5362114/config.verbose.txt)
[config.info.txt](https://github.com/HandBrake/HandBrake/files/5362115/config.info.txt)
[build.txt](https://github.com/HandBrake/HandBrake/files/5362116/build.txt)

Looks like it started into the build process though.

Nomis101 commented:
> can you post the whole arm64 xcodebuild log? Which autoconf version are you using?

autoconf (GNU Autoconf) 2.69

It is not the complete build log, because it does not work anymore. I don't know why, I did not change anything.
[arm64-output.txt](https://github.com/HandBrake/HandBrake/files/5362242/arm64-output.txt)

Nomis101 commented:
One issue seems to be, after `make ub.build.serial` the `HOST.arch` in /build/ub.osx1100.arm64/GNUmakefile is set to x86_64 instead of arm64. Then the make process tries to build two x86_64 versions.


> compute: makevar BUILD/  = .
> compute: makevar PREFIX/ = /usr/local
> compute: makevar XCODE.prefix/ = /Applications
> mkdir: ub.osx1100.**arm64**
> chdir: ub.osx1100.**arm64**
> write: GNUmakefile
> mkdir: project
> write: project/handbrake.m4
> write: distfile.cfg
> Build system:       **x86_64**-apple-darwin20.1.0
> Host system:        **x86_64**-apple-darwin20.1.0
> Target platform:    darwin

It should be
> Build system:       **x86_64**-apple-darwin20.1.0
> Host system:        **arm64**-apple-darwin20.1.0

galad87 commented:
That shouldn't be relevant, because it will call xcodebuild with the arm64 config and ignore those settings.
Do you have both Xcode 12 or earlier and Xcode 12.2 beta on the disk or only the 12.2 beta?

Nomis101 commented:
I have only the 12.2 beta.

bradleysepos commented:
Super clean macOS 10.15.7 with Xcode 12.2 beta 3. Toolchain installed via `scripts/mac-toolchain-build`, version 2.5.0.

arm64, via `--arch`, cli-only:

```
$ ./configure --sandbox --harden --arch=arm64 --disable-xcode
$ cd build
$ make -j1
checking host system type... Invalid configuration `arm64-apple-darwin19.6.0': machine `arm64-apple' not recognized
configure: error: /bin/sh ./config.sub arm64-apple-darwin19.6.0 failed
make: *** [contrib/libopus/.stamp.libopus.configure] Error 1
```

arm64 via `--arch`, no `--xcode-config`:

```
$ ./configure --sandbox --harden --arch=arm64
$ cd build
$ make -j1
# builds x86_64
```

arm64 via `--xcode-config`:

```
$ ./configure --sandbox --harden --xcode-config=osx1011.arm64
$ cd build
$ make -j1
compute: available architectures...(pass) arm64 x86_64 [x86_64]
Build system:       x86_64-apple-darwin19.6.0
Host system:        x86_64-apple-darwin19.6.0
checking host system type... Invalid configuration `arm64-apple-darwin19.6.0': machine `arm64-apple' not recognized
configure: error: /bin/sh ./config.sub arm64-apple-darwin19.6.0 failed
make[1]: *** [contrib/libopus/.stamp.libopus.configure] Error 1
```

I'll look into this a bit.

bradleysepos commented:
Some potential fixes on my branch: https://github.com/bradleysepos/HandBrake/commits/applecross-test, specifically https://github.com/bradleysepos/HandBrake/commit/2dc299537e039cbd927675875d5c77cfd512384a and https://github.com/bradleysepos/HandBrake/commit/8dd5dee6bb807f0c4f8f1ed257b64efa1740e1d9.

libopus should build on 10.15.7 with Xcode 12.2 beta (3) now. x264 fails on aarch64 asm but I'll have to look into that later (others feel free, I'm just out of time for now).

Nomis101 commented:
Yes, build process is different now. Now for me it stops with:

```
checking for suffix of executables... 
checking whether we are cross compiling... configure: error: in `/Volumes/Developer/UB/HandBrake/build/contrib/lame/lame-3.100':
configure: error: cannot run C compiled programs.
If you meant to cross compile, use `--host'.
See `config.log' for more details
make: *** [contrib/lame/.stamp.lame.configure] Error 1
```

or

```
checking for suffix of executables... 
checking whether we are cross compiling... configure: error: in `/Volumes/Developer/UB/HandBrake/build/contrib/libogg/libogg-1.3.4':
configure: error: cannot run C compiled programs.
If you meant to cross compile, use `--host'.
See `config.log' for more details
make: *** [contrib/libogg/.stamp.libogg.configure] Error 1
```

If doing `../configure --disable-xcode --arch=arm64 && make build`

But this can be easy fixed by adding `--host` to module.defs

galad87 commented:
On arm macOS the difference between build_tuple.spec (aarch64-apple-darwin20.1.0) and build_tuple.machine (arm64) creates some monsters and enables cross-compiling from arm to arm:

```
Build system:       aarch64-apple-darwin20.1.0
Host system:        arm64ch64-apple-darwin20.1.0
Target platform:    darwin (cross-compile)
```

Cross compiling from arm to intel worked.

bradleysepos commented:
Ah, I'm under the impression the tuple is supposed to be arm64-apple-darwin\* and arch arm64, code in `configure.py` tries to make it so. Something's definitely wrong if you're seeing arm64ch64 (wut), so I'll try to sort that.

It seems clear we need arch to be arm64 but is the spec supposed to be aarch64-\*? GNU config.guess identifies as aarch64-* but then Apple stuffs are like what is aarch64 so this continues to be fun.

Nomis101 commented:
Yes, the different naming schemes confuses me. Clang doesn't like `aarch64-apple-darwin`, the triple needs to be `arm64-apple-darwin` for clang. But others (config.guess, rust ...) are using `aarch64-apple-darwin`. Very confusing.

bradleysepos commented:
Indeed. I think what needs to happen is we let `config.guess` identify as aarch64, modify to arm64 for apple/clang, and try to use aarch64 for all contribs unless otherwise indicated. I made the latter happen for libopus via make substitution (`$(HOST.machine:arm64=aarch64)`) but perhaps I should look into doing that higher up in the build system.

galad87 commented:
@bradleysepos with your configure.py changes HOST.cross is not defined to 1 when cross-compiling on Mac. My workaround is:

`    doc.add( 'HOST.cross', int(options.cross != None or arch.mode.mode != arch.mode.default or build_tuple.machine != host_tuple.machine) )
`

in line 2014. Probably there is a better place.

galad87 commented:
Cross-compiling from intel to arm completed successfully on the GitHub action, please try again to compile on macOS 11 and Xcode 12.2.

Nomis101 commented:
This is great, it works now! 😀 I can now cross-compile HandBrakeCLI to arm64 with `mkdir build && cd build && ../configure --disable-xcode --arch=arm64 && make build`. I've checked the architecture and `lipo -archs` tells me arm64.

Universal binary and a GUI build still doesn't work, but be able to cross-compile to arm64 is a big step. 👍🏻

galad87 commented:
That's caused by either the Sparkle submodule not being inited, or Xcode caching some weird things. Try to clear the ~/Library/Developer/Xcode/Derived Data folder.

Nomis101 commented:
You are right, after deleting the Derived Data folder, the GUI build also works. I cross-compiled an arm64 HandBrake GUI successfully. Of course I don't know if the build can be started, because I don't have an Apple Silicon Mac but the Info dialog tells me its an Apple Silicon application.

Nomis101 commented:
Universal binary also works after deleting the Derived Data folder. 🎉

```
lipo -archs /Volumes/Developer/UB/HandBrake/build/xroot/HandBrakeCLI
x86_64 arm64
```

galad87 committed:
contrib: macOS arm and intel cross compilation.

galad87 committed:
make: Improve setting host system in configure.py where arch or xcode-config is set.

Assists building for arm64-apple-darwin (Apple Silicon) on x86_64-apple-darwin.

galad87 committed:
build: improve ub target to generate a arm64/x86_64 binary.

galad87 committed:
actions: enable universal mac build.

galad87 committed:
build: add 'MACOSX' prefix to the 'UB' defines.

sr55 commented:
Everything is building and packaging fine for me now. Thanks!

bradleysepos commented on .github/workflows/mac.yml:
```diff
@@ -35,8 +38,10 @@ jobs:
-        ./configure --launch-jobs=$(sysctl -n hw.ncpu) --launch
+        mkdir build
         cd build
+        ../configure --sandbox --harden
```
Move up one line (before `cd build`) and remove leading dot (`./configure ...`)?

bradleysepos commented on contrib/libdav1d/module.defs:
```diff
@@ -33,6 +33,10 @@ ifeq (arm64-darwin,$(HOST.machine)-$(HOST.system))
     LIBDAV1D.CONFIGURE.extra += --cross-file=$(call fn.ABSOLUTE,$(LIBDAV1D.EXTRACT.dir/))/arm64-darwin-gcc.meson -Dc_args="-arch arm64"
 endif
 
+ifeq (x86_64-darwin,$(HOST.machine)-$(HOST.system))
```
I would probably combine these into check darwin, check arch inside (nested). Cosmetic.

bradleysepos commented on contrib/libjpeg-turbo/module.defs:
```diff
@@ -26,6 +26,13 @@ else
+ifeq (arm64-darwin,$(HOST.machine)-$(HOST.system))
+    LIBJPEGTURBO.CONFIGURE.extra += -DCMAKE_SYSTEM_PROCESSOR=arm64
+endif
+ifeq (x86_64-darwin,$(HOST.machine)-$(HOST.system))
```
Combine with previous ifeq (cosmetic)?

bradleysepos commented on contrib/libopus/module.defs:
```diff
@@ -12,4 +12,11 @@ LIBOPUS.CONFIGURE.extra = --disable-doc --disable-extra-programs
+    endif
+endif
+
+LIBOPUS.CONFIGURE.bootstrap = rm -fr aclocal.m4 autom4te.cache configure; autoreconf -fiv;
```
LF

bradleysepos commented on contrib/libtheora/module.defs:
```diff
@@ -20,3 +20,10 @@ LIBTHEORA.CONFIGURE.extra = \
+        LIBTHEORA.CONFIGURE.args.host = --host=$(HOST.machine:arm64=aarch64)-$(HOST.vendor)-$(HOST.system)
+        LIBTHEORA.CONFIGURE.args.build = --build=$(BUILD.machine:arm64=aarch64)-$(BUILD.vendor)-$(BUILD.system)
+    endif
+endif
```
LF

bradleysepos commented on contrib/libvorbis/module.defs:
```diff
@@ -8,3 +8,10 @@ LIBVORBIS.FETCH.sha256  = 6ed40e0241089a42c48604dc00e362beee00036af2d8b3f4633803
+        LIBVORBIS.CONFIGURE.args.host = --host=$(HOST.machine:arm64=aarch64)-$(HOST.vendor)-$(HOST.system)
+        LIBVORBIS.CONFIGURE.args.build = --build=$(BUILD.machine:arm64=aarch64)-$(BUILD.vendor)-$(BUILD.system)
+    endif
+endif
```
LF

bradleysepos commented on contrib/x264/module.defs:
```diff
@@ -14,9 +14,14 @@ X264.CONFIGURE.extra += --disable-lavf --disable-ffms --disable-avs --disable-sw
+    endif
     X264.CONFIGURE.env.CFLAGS = CFLAGS="-I$(call fn.ABSOLUTE,$(CONTRIB.build/)include) $(call fn.ARGS,X264.GCC,*archs *sysroot *minver ?extra)"
     X264.CONFIGURE.env.LDFLAGS = LDFLAGS="-L$(call fn.ABSOLUTE,$(CONTRIB.build/)lib) $(call fn.ARGS,X264.GCC,*archs *sysroot *minver)"
+    ifeq (arm64-darwin,$(HOST.machine)-$(HOST.system))
```
Do we not need to do this for cross compiling arm64->x86_64? Same for lines 27-29.

galad87 replied:
It doesn't seem needed. I'll recheck it soon.

bradleysepos commented on contrib/x265_10bit/module.defs:
```diff
@@ -29,6 +29,9 @@ endif
 ifeq (arm64-darwin,$(HOST.machine)-$(HOST.system))
         X265_10.CONFIGURE.extra += -DCMAKE_OSX_ARCHITECTURES=arm64
 endif
+ifeq (x86_64-darwin,$(HOST.machine)-$(HOST.system))
```
Combine with previous ifeq (cosmetic)?

bradleysepos commented on contrib/x265_8bit/module.defs:
```diff
@@ -27,6 +27,10 @@ ifeq (arm64-darwin,$(HOST.machine)-$(HOST.system))
         X265_8.CONFIGURE.extra += -DCMAKE_OSX_ARCHITECTURES=arm64
 endif
 
+ifeq (x86_64-darwin,$(HOST.machine)-$(HOST.system))
```
Combine with previous ifeq (cosmetic)?

bradleysepos commented on make/configure.py:
```diff
@@ -2185,4 +2207,4 @@ class Tools:
     sys.exit( 1 )
 
-sys.exit( 0 )
+sys.exit( 0 )
```
LF

bradleysepos reviewed (approved):
Approved with a few comments, mostly cosmetic. Can land as-is if time demands.

galad87 committed:
contrib: combine some checks and cosmetics.

galad87 committed:
actions: simplify mac workflow.

bradleysepos commented:
Thanks everyone, especially @Nomis101 for being a superb guinea pig (test subject)! 🐹 🚀
